src/kernels/MIOpenCosineEmbeddingLoss.cpp

```diff
@@ -0,0 +1,462 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+#ifndef INPUT_TYPE
+#define INPUT_TYPE float
+#endif
+
+#ifndef OUTPUT_TYPE
+#define OUTPUT_TYPE float
+#endif
+
+#ifndef D_TYPE
+#define D_TYPE float
+#endif
+
+#ifndef REDUCE_SIZE
+#define REDUCE_SIZE 256
+#endif
+
+__device__ FLOAT_ACCUM warpReduceSum(FLOAT_ACCUM val)
+{
+    if(warpSize >= 64)
+        val += __shfl_down(val, 32);
+    if(warpSize >= 32)
+        val += __shfl_down(val, 16);
+    if(warpSize >= 16)
+        val += __shfl_down(val, 8);
+    if(warpSize >= 8)
+        val += __shfl_down(val, 4);
+    if(warpSize >= 4)
+        val += __shfl_down(val, 2);
+    if(warpSize >= 2)
+        val += __shfl_down(val, 1);
+    return val;
+}
+
+__device__ FLOAT_ACCUM blockReduceSum(FLOAT_ACCUM val)
+{
+    static __shared__ FLOAT_ACCUM shared[REDUCE_SIZE / warpSize];
+    auto lane = threadIdx.x % warpSize;
+    auto wid  = threadIdx.x / warpSize;
+
+    val = warpReduceSum(val);
+
+    if(lane == 0)
+        shared[wid] = val;
+    __syncthreads();
+
+    val = threadIdx.x < REDUCE_SIZE / warpSize ? shared[lane] : CVT_FP32_2ACCUM(0.0f);
+    if(wid == 0)
+        val = warpReduceSum(val);
+
+    return val;
+}
+
+template <typename DTYPE>
+__device__ void lossSum(const DTYPE* input, DTYPE* output, size_t N)
+{
+    auto gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    FLOAT_ACCUM val = gid < N ? CVT_FLOAT2ACCUM(input[gid]) : CVT_FP32_2ACCUM(0.0f);
+    val             = blockReduceSum(val);
+
+    if(threadIdx.x == 0)
+    {
+        output[blockIdx.x] = CVT_ACCUM2FLOAT(val);
+    }
+}
+
+extern "C" __global__ void
+LossSum(const D_TYPE* __restrict__ input, D_TYPE* __restrict__ output, size_t N)
```

Comment:
These parts overlap with other PRs. Please refer to https://github.com/moreh-dev/MIOpen/pull/20 and separate it into a separate file.

Reply:
separated it to MIOpenLossSum.cpp